src/payment/strategies/mollie/mollie-payment-strategy.ts

```diff
@@ -71,16 +85,29 @@ export default class MolliePaymentStrategy implements PaymentStrategy {
         const shouldSaveInstrument = (paymentData as HostedInstrument)?.shouldSaveInstrument;
         const shouldSetAsDefaultInstrument = (paymentData as HostedInstrument)?.shouldSetAsDefaultInstrument;
 
-        if (!payment || !paymentData) {
-            throw new PaymentArgumentInvalidError([ 'payment' ]);
+        if (!payment || !payment.gatewayId || !paymentData) {
+            throw new PaymentArgumentInvalidError([ 'payment', 'gatewayId', 'paymentData' ]);
         }
 
         try {
             if (payment.methodId === MolliePaymentMethodType.creditcard) {
                 await this._store.dispatch(this._orderActionCreator.submitOrder(order, options));
 
                 if (paymentData && isVaultedInstrument(paymentData)) {
-                    return this._store.dispatch(this._paymentActionCreator.submitPayment(payment));
+                    if (this._isHostedPaymentFormEnabled(payment.methodId, payment.gatewayId) && this._shouldRenderHostedForm) {
+                        const form = this._hostedForm;
+
+                        if (!form) {
+                            throw new NotInitializedError(NotInitializedErrorType.PaymentNotInitialized);
+                        }
+
+                        await form.validate();
+                        await form.submit(payment);
+
+                        return await this._store.dispatch(this._orderActionCreator.loadCurrentOrder());
+                    } else {
+                        return await this._store.dispatch(this._paymentActionCreator.submitPayment(payment));
+                    }
```

Comment:
🍹 This might be a good point to split the different paths into different `execute` methods. I can see we have at least 4 different paths in here: `executeCreditCard`/`executeWithInstrumentHosted`/`executeOther`/`executeWithInstrument`. What do you think?

That way this method could just route to the actual method performing the action:

```ts
try {
	if (payment.methodId !== MolliePaymentMethodType.creditcard) {
		return this._executeOther();
    }

	if (!paymentData || !isVaultedInstrument(paymentData) {
		return this._executeCreditCard();
	}

	if (this._isHostedPaymentFormEnabled(payment.methodId, payment.gatewayId) && this._shouldRenderHostedForm) {
		return this._executeWithInstrumentHosted();
    }

    return this._executeWithInstrument();
} catch (error) {
	// rest of the code...
}
```

Reply:
Can you review my approach, i did it i bit different let me know what you think